src/crimson/net/Protocol.h

```diff
@@ -0,0 +1,125 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2018 Intel Corp.
+ *
+ * Author: Yingxin Cheng <yingxincheng@gmail.com>
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#pragma once
+
+#include "Fwd.h"
+#include "Session.h"
+#include "Messenger.h"
+
+#include <seastar/core/shared_future.hh>
+
+namespace ceph::net {
+
+enum class state_t {
+  none,
+  accept,
+  connect,
+  open,
+  // standby,
+  // wait,
+  close,
+  LAST
+};
+
+class Protocol
```

Comment:
what's the advantage of break `SocketConnection` down into multiple `Protocol` classes? what i found is mostly the state transition logic here. could you elaborate a little bit on how it helps with implementing different policies and msgr v2?

Reply:
explained in https://github.com/ceph/ceph/pull/24142#issuecomment-425357280